src/array.jl

```diff
@@ -767,41 +767,52 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
                      :levels!)
         allowmissing = allow_missing
     end
-    if !allunique(newlevels)
-        throw(ArgumentError(string("duplicated levels found: ",
-                                   join(unique(filter(x->sum(newlevels.==x)>1, newlevels)), ", "))))
+    (levels(A) == newlevels) && return A # nothing to do
+
+    # map each new level to its ref code and number of occurrences
+    newlv2ref = Dict{eltype(newlevels), Pair{Int, Int}}()
+    dupnewlvs = similar(newlevels, 0)
+    for (i, lv) in enumerate(newlevels)
+        lvref = get(newlv2ref, lv, i => 0)
+        newlv2ref[lv] = lvref[1] => lvref[2] + 1
+        # remember the first duplicate occurrence
+        (lvref[2] == 1) && push!(dupnewlvs, lv)
+    end
+    if !isempty(dupnewlvs)
+        throw(ArgumentError(string("duplicated levels found: ", join(dupnewlvs, ", "))))
     end
 
-    oldlevels = levels(A.pool)
+    # map each old ref code to new ref code (or 0 if no such level)
+    oldlevels = levels(pool(A))
+    oldref2newref = fill(0, length(oldlevels) + 1)
+    for (i, lv) in enumerate(oldlevels)
+        oldref2newref[i + 1] = first(get(newlv2ref, lv, 0 => 0))
+    end
 
     # first pass to check whether, if some levels are removed, changes can be applied without error
-    # TODO: save original levels and undo changes in case of error to skip this step
-    # equivalent to issubset but faster due to JuliaLang/julia#24624
```

Comment:
The TODO still applies AFAICT. The idea was that we could avoid this first pass and instead do the check in the recoding loop, if we had some code to revert the changes made up to the problematic entry when an error happens.

Reply:
I can keep it, but wouldn't always saving the old refs be more memory expensive + saving itself _is_ a pass that affects the cache?

Reply:
Also, before this PR the checks were more expensive as `oldref2newref` was not available, but now the total cost of the pass is *O(N)* and it only involves integer arrays.

Reply:
Well, potentially saving the old `A.refs` could be avoid as until the missing level has encountered the changes should be reversible (with `newref2oldref`).
I will commit this variant, you can then decide whether it's worth keeping.

Reply:
Thanks, that's what I had in mind. Can you add tests to ensure that recoding is correct in the various error cases?

Reply:
I've added some tests (helped to fix one bug).